ext/gvl_tracing_native_extension/gvl_tracing.c

```diff
@@ -42,54 +42,131 @@
-static _Thread_local bool sleeping = false; // Used to track when a thread is sleeping
+static const char* get_event_name(rb_event_flag_t event_id);
 
 // Global mutable state
-static rb_atomic_t thread_serial = 0;
 static FILE *output_file = NULL;
 static rb_internal_thread_event_hook_t *current_hook = NULL;
 static double started_tracing_at_microseconds = 0;
 static int64_t process_id = 0;
 static VALUE gc_tracepoint = Qnil;
 
+#if RUBY_3_3_PLUS
+static int gvl_tracing_key = 0;
+#else
+static _Thread_local thread_meta_t thread_local_meta = {};
+static rb_atomic_t thread_serial = 0;
+static void set_native_thread_id(void);
+#endif
+
 void Init_gvl_tracing_native_extension(void) {
   rb_global_variable(&gc_tracepoint);
 
+#if RUBY_3_3_PLUS
+  gvl_tracing_key = rb_internal_thread_specific_key_create();
+#endif
   VALUE gvl_tracing_module = rb_define_module("GvlTracing");
 
   rb_define_singleton_method(gvl_tracing_module, "_start", tracing_start, 1);
   rb_define_singleton_method(gvl_tracing_module, "_stop", tracing_stop, 0);
+  rb_define_singleton_method(gvl_tracing_module, "_clean", tracing_clean, 1);
+  rb_define_singleton_method(gvl_tracing_module, "_use_object_id?", use_object_id, 0);
   rb_define_singleton_method(gvl_tracing_module, "mark_sleeping", mark_sleeping, 0);
 }
 
-static inline void initialize_thread_id(void) {
-  current_thread_seen = true;
-  current_thread_serial = RUBY_ATOMIC_FETCH_ADD(thread_serial, 1);
+#if RUBY_3_3_PLUS
+
+static thread_meta_t* current_thread_meta(VALUE thread) {
+  void *data = rb_internal_thread_specific_get(thread, gvl_tracing_key);
+  thread_meta_t* thread_meta = NULL;
+
+  if (data != NULL) {
+    thread_meta = (thread_meta_t *)data;
+  } else {
+    thread_meta = (thread_meta_t*)malloc(sizeof(thread_meta_t));
+    memset(thread_meta, 0, sizeof(thread_meta_t));
+    thread_meta->thread = thread;
+    thread_meta->thread_object_id = RB_NUM2INT(rb_obj_id(thread));
+    rb_internal_thread_specific_set(thread, gvl_tracing_key, (void *)thread_meta);
+  }
+
+  return thread_meta;
+}
+
+static inline void initialize_thread_id(thread_meta_t *thread_meta) {
+  thread_meta->current_thread_seen = true;
+  thread_meta->thread_object_id = RB_NUM2INT(rb_obj_id(thread_meta->thread));
+}
+
+static void free_thread_meta(VALUE thread) {
+  void *data = rb_internal_thread_specific_get(thread, gvl_tracing_key);
+  if (data != NULL) {
+    free(data);
+    rb_internal_thread_specific_set(thread, gvl_tracing_key, (void *)NULL);
+  }
+}
+
+#else
+
+static thread_meta_t* current_thread_meta(VALUE thread) {
+  return &thread_local_meta;
+}
+
+static void set_native_thread_id(void) {
+  uint64_t native_thread_id = 0;
+
+  #ifdef HAVE_PTHREAD_THREADID_NP
+    pthread_threadid_np(pthread_self(), &native_thread_id);
+  #elif HAVE_GETTID
+    native_thread_id = gettid();
+  #else
+    native_thread_id = thread_local_meta->current_thread_serial; // TODO: Better fallback for Windows?
```

Comment:
Not sure if you know, but the GVL API is a noop on Windows

Reply:
Uh, I didn't know about that!

In that case, maybe go with removing `current_thread_serial` and adding an `ifdef` to make the gem become a no-op on windows may be the way to go.

(I think it's still nice if the gem builds/installs, so that folks don't need to have conditions on their `Gemfile`s)